lock/common.go

```diff
@@ -0,0 +1,32 @@
+// Package lock provides implementations for cancellable and distributed locks.
+package lock
+
+import (
+	"golang.org/x/net/context"
+	"gopkg.in/src-d/go-errors.v0"
+)
+
+var (
+	ErrCanceled      = errors.NewKind("context canceled")
+	ErrAlreadyClosed = errors.NewKind("already closed")
+)
+
+// Locker is a service providing locks.
+type Locker interface {
+	// Lock acquires a lock with the given id.
+	// Context can be used to cancel the operation (e.g. with a timeout).
+	// If it succeeds, it returns a Lock. Otherwise, it returns an error.
+	Lock(id string, ctx context.Context) (Lock, error)
+	// Close closes the locker and releases any resources held by it.
+	// If it is called more than once, ErrAlreadyClosed is returned.
+	Close() error
+}
+
+// Lock is an acquired lock.
+type Lock interface {
+	// Unlock releases the lock.
+	Unlock()
```

Comment:
Implementations need to make sure that if they cannot release the lock, it will expire.